Only rename method type params shadowing enclosing class type params

tanishiking opened the pull request:
Previously, method type parameters were always renamed if they had the same name as any class type parameter, regardless of whether the class type parameter was actually used in the method signature.

As a result, we rename methods's type params for unnecessary names like: https://github.com/scala/scala3/issues/24671 (Note that renaming actually doesn't cause binary incompatibility, and this change is just for make generic signature a bit clear and silence false-positive MIMA reporting).

For example, in an example below, the Scala compiler rename the generic signature of `bar` to something like `bar[T1](x: T1): T1` because `T` is used by `Foo[T]`. However, this is unnessary rename because none of T in method signature refer the `T` of `Foo[T]`.

```scala
class Foo[T]:
  def bar[T](x: T): T = ???
```

This commit makes the renaming conditional:
Method type parameters are only renamed when
- (1) A class type parameter is referenced in the method signature, and
- (2) That class type parameter is shadowed by a method type parameter

tanishiking commented on tests/generic-java-signatures/shadowedTypeParam.scala:
```diff
@@ -0,0 +1,43 @@
+  // copy$default$1 still have `<T1> T Bar.copy$default$1` rather than `<T> T Bar.copy$default$1`
+  // as reported in https://github.com/scala/scala3/issues/24671
+  // The type parameter rename occurs because the return type T refers the enclosing class's type param T.
+  // printMethodSig(barMethods, "copy$default$1")
```
Still MiMa won't be happy with `<T1> T Bar.copy$default$1` rather than `<T> T Bar.copy$default$1` :(

lrytz commented on compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:
```diff
@@ -528,4 +546,27 @@ object GenericSignatures {
+        case _ =>
+          traverseChildren(tp)
+
+    types.foreach(collector.traverse)
```
```suggestion
    def collector(tp: Type): Unit = tp.foreachPart:
      case ref@TypeParamRef(_: PolyType, _) =>
        usedMethodTypeParamNames += ref.paramName
      case TypeRef(pre, _) =>
        val sym = tp.typeSymbol
        if isTypeParameterInMethSig(sym, initialSymbol) then
          usedMethodTypeParamNames += sym.name
        else if sym.isTypeParam && sym.isContainedIn(initialSymbol.topLevelClass) then
          usedClassTypeParams += sym
      case _ =>

    types.foreach(collector)
```

lrytz commented on compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:
```diff
@@ -528,4 +546,27 @@ object GenericSignatures {
+          else if sym.isTypeParam && sym.isContainedIn(initialSymbol.topLevelClass) then
+            usedClassTypeParams += sym
+          else
+            traverse(pre)
```
this should probably not be in the `else` branch?

tanishiking replied:
Right, we don't need to traverse the `pre`

lrytz commented on compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:
```diff
@@ -448,6 +460,12 @@ object GenericSignatures {
         (initialSymbol.is(Method) && initialSymbol.typeParams.contains(sym))
       )
 
+  private def isTypeParameterInMethSig(sym: Symbol, initialSymbol: Symbol)(using Context) =
```
I would probably nest this method within `collectUsedTypeParams`

tanishiking commented:
Thanks @lrytz ! Could you take an another look? (Added a change based on your suggestion)

lrytz commented on compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala:
```diff
@@ -528,4 +546,24 @@ object GenericSignatures {
+      case tp: TypeRef =>
+        val sym = tp.typeSymbol
+        if isTypeParameterInMethSig(sym, initialSymbol) then
+          usedMethodTypeParamNames += sym.name
```
do we even need this case, aren't all references to the method's type params handled above?

tanishiking commented:
Thanks @lrytz, refactored the `collectUsedTypeParams` based on your feedback, and added test case for https://github.com/scala/scala3/issues/24619, I'll merge this PR once CI passed.

lrytz reviewed (approved):
👌
